### DISCUSSION TITLE
How can I test if secrets are available in an action?

### DISCUSSION THREAD

**Discussion by paulb777** [community/community#26726] [Actions]
<p>I’d like an action to run on PRs from the main repo, but not from PRs from forked repos.</p>

**Comment by Yanjingzhu** [comment]
<p>Secrets are not passed to workflows that are triggered by a pull request from a fork. As you said , your PRs are within the main repo. Then the secrets could be passed in workflow. </p>
<p>How do you use secrets in your action? Use secret variables as action input variable value? </p>
<p><img src="https://user-images.githubusercontent.com/24806478/181099432-54114c3f-fe5a-49f5-b3c1-5ba76bfd7728.png" alt="use secrets.png" data-base62-sha1="8TEY28DGwxUxiU3NmJE1sg6wVlB" width="423" height="389"> </p>
<p>If so, you could check your logs , the secrets variable value will be masked with *** . </p>
<p><img src="https://user-images.githubusercontent.com/24806478/181099433-fc0ee827-b527-41fd-8e8c-3ef631f11339.png" alt="secrets.png" data-base62-sha1="7gAwOEUUTOYPhswf7xiSSfyfiao" width="413" height="288"></p>
<p>If I misunderstanding your scenario, please share your workflow yml content here.  </p>

**Comment by paulb777** [comment]
<p>To clarify, I’d like my CI to run all the jobs and pass when run from the main repo. When someone sends a PR from a forked repo, I’d still like the subset of jobs that don’t require secrets to run and CI still to pass. </p>
<p>I’ll follow up with a detailed example in the next day or two…</p>

**Comment by paulb777** [comment]
<p>I was able to get it working by testing every step of the job for the existence of the environment variable associated with the secret.  See <a href="https://github.com/firebase/firebase-ios-sdk/pull/5180" rel="nofollow noopener">https://github.com/firebase/firebase-ios-sdk/pull/5180</a>.</p>
<p>It would be nicer if there were away to check for secrets availability at the job level.</p>

**Comment by Yanjingzhu** [comment]
<p><a class="mention" href="/paulb777">@paulb777</a> I checked your PR, you use secrets as the value of environment variables. You could set the env in job level.  Then the env could be used in your scripts directly. In bash, use it in syntax $var_name</p>
<pre><code>jobs:
  build:
    env:
      key1: ${{secrets.test2}}
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - run: echo $key1
      if: ${{env.key1}} == 'aaa'
</code></pre>
<p>You could enable <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#enabling-step-debug-logging" rel="nofollow noopener">step debug logging</a>, in set up job step, the secrets will be evaluated. </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/f4f68767f442ae6b87ae1fe2facb0e4e3e7c28dd.png" data-download-href="/uploads/short-url/yX2GGVkQJeBkey5JmXFJagEYCQd.png?dl=1" title="set up job.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181099437-568fa09d-9a01-4baa-9cb3-49ddfcbd0890.png" alt="set up job.png" data-base62-sha1="yX2GGVkQJeBkey5JmXFJagEYCQd" width="690" height="323" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/f4f68767f442ae6b87ae1fe2facb0e4e3e7c28dd_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">set up job.png</span><span class="informations">862×404 10.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by paulb777** [comment]
<p>Thanks <a class="mention" href="/yanjingzhu">@yanjingzhu</a>. Setting the <a href="https://github.com/firebase/firebase-ios-sdk/pull/5188" rel="nofollow noopener">secret environment variables at the job level</a> is much cleaner.</p>
<p>I tried several different locations and variations for the <code>if</code> (see the commits in <a href="https://github.com/firebase/firebase-ios-sdk/pull/5188" rel="nofollow noopener">https://github.com/firebase/firebase-ios-sdk/pull/5188)</a>, but wasn’t able to find a way to use it to disable the job when the secret environment variables are not available. </p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>You can only use the <strong>env</strong> context in the value of the <strong>with</strong> and <strong>name</strong> keys, or in a <strong>step’s if</strong> conditional.</p>
<p>It is not supported to use <strong>env</strong> in <strong>job’s if</strong>  conditional. </p>
<p>And screrts context could not be used in if conditional, neither <strong>job’s</strong>  <strong>if</strong>  nor  <strong>step’s if</strong>  .</p>
<p>So, it is not possible to disable a job by identifying secrets . I am afraid that you need to add if contional to each steps.  Sorry for any inconvenience. </p>

**Reply by dwertent** [reply]
Thank you for the detailed reply.
This is really inconvenient, is there any plan on changing this?

**Reply by itsalaidbacklife** [reply]
I would also love to see an official approach to handling this. In particular, this is something that would be highly useful for open source repositories where some updates are managed through forked clones and some through branches in the original repository. Workflows triggered by forks will not have access to secrets and so being able to neatly handle those in a different way would be a very helpful.

**Comment by ojeytonwilliams** [comment]
<p>This might not be what you’re after, but you can disable individual steps by putting the secret into a job’s env and then using</p>
<p>if: ${{ env.SECRET_KEY != 0 }}<br>
inside each step.</p>
<p>Setting env at the workflow level does not appear to work, so ‘if’ for entire jobs won’t work.</p>

**Comment by jonico** [comment]
<p>You can make an entire job optional based on the presence of a secret if you had a previous job that set a job output: <a href="https://gist.github.com/jonico/24ffebee6d2fa2e679389fac8aef50a3" rel="nofollow noopener">https://gist.github.com/jonico/24ffebee6d2fa2e679389fac8aef50a3</a></p>

**Comment by GabLeRoux** [comment]
Thanks jonico, I’ve created a more specific example here:  
https://github.com/GabLeRoux/github-actions-examples/blob/e0468ce2731b08bd8b1f7cd09d0b94c541310693/.github/workflows/secret_based_conditions.yml

**Comment by TomerFi** [comment]
<p>This did the trick for me:</p>
<pre><code class="lang-yaml">- name: Clean local maven repository
  env:
    CLEAN_MVN_REPO: ${{ secrets.CLEAN_MVN_REPO }}
  if: env.CLEAN_MVN_REPO != null
  run: rm -r ~/.m2/repository/*
</code></pre>
<p>If the <code>secret</code> called <code>CLEAN_MVN_REPO </code> doesn’t exist,<br>
this step isn’t executed.</p>

**Comment by stephan-rayner** [comment]
<p>It may be a little cheeky but this is what I did at my org. I run a job before my main CI job that checks if all my secrets are in place. It normally does other things too but I took that out for the example below.</p>
<pre><code class="lang-auto">...

jobs:

  preFlight:
    name: "Preflight Checks"
    runs-on: ubuntu-latest
    steps:

    - name: Assert Secrets Present
      shell: bash
      run: |
        MD5() {
          result=($(md5sum &lt;(echo "${1}")))
          echo $result
        }

        secretPresent() {
          emptyHash=$(MD5 "")
          secretHash=$(MD5 "${1}")
          ! diff &lt;(echo $secretHash) &lt;(echo $emptyHash) &amp;&gt; /dev/null
        }

        errorMessage() {
          echo "Your secret ${1} seems to be missing please contact the Central-OPs team"
          return 1
        }

        secretStatus=0

        secretPresent "${{ secrets.GIT_TOKEN }}" || errorMessage "GIT_TOKEN" || secretStatus=1
        secretPresent "${{ secrets.AWD_KEY }}" || errorMessage "AWS_KEY" || secretStatus=2
        secretPresent "${{ secrets.THING_URL }}" || errorMessage "THING_URL" || secretStatus=3
        secretPresent "${{ secrets.THING_USER_TOKEN }}" || errorMessage "THING_USER_TOKEN" || secretStatus=4
        secretPresent "${{ secrets.CHEESE }}" || errorMessage "CHEESE" || secretStatus=5

        exit $secretStatus

    - run: echo "Everything Checks Out 👍"

  build:
    needs: preFlight
    name: "Build and Push to Registry"
    runs-on: ubuntu-latest
    steps:
    
    - name: Checkout Code Base
      uses: actions/checkout@v2

    ...
</code></pre>
<p>That said I also really like the solution by <a href="https://github.community/t/how-can-i-test-if-secrets-are-available-in-an-action/17911/11">tomerfi</a>.</p>

**Comment by lbergelson** [comment]
<p>Thank you all for the workarounds, but this is ridiculous, there should be a top level github.secrets_present variable or something similar.  I don’t understand why these extremely basic issues haven’t been addressed by github.</p>

**Comment by swissbuechi** [comment]
I solved it the following way:
```yaml
...
      - name: myjob
        env:
          MY_SECRET: ${{ secrets.MY_SECRET }}
        if: env.MY_SECRET != null
        run: echo "running"
...
```
Credits to @TomerFi and @ojeytonwilliams 

Do you guys have any security concerns? Is it save to expose a secret as a local environment variable during a step?

**Reply by TomerFi** [reply]
I think GitHub consilds the secrets content in logs and outputs.
Other then that, it's quite a common technique I think, especially for actions that take their arguments via environment variables and not via the action's spec (the _with_ directive).

**Comment by MikeMcC399** [comment]
https://stackoverflow.com/questions/70249519/how-to-check-if-a-secret-variable-is-empty-in-if-conditional-github-actions also suggests a workaround for this issue.

[GH Actions: Context](https://docs.github.com/en/actions/learn-github-actions/contexts#context-availability) backs up what @Yanjingzhu wrote.

**Reply by tomas-pajurek** [reply]
I have recently also struggled with this and came up with the solution involving `github.secret_source` property. I am explaining this in the StackOverflow question you referenced.
